# **(Part 4 Feature Selection And Final Model)**

## Objectives

* Write here your notebook objective, for example, "Fetch data from Kaggle and save as raw data", or "engineer features for modelling"

## Inputs

* Write here which data or information you need to run the notebook 

## Outputs

* Write here which files, code or artefacts you generate by the end of the notebook 

## Additional Comments

* In case you have any additional comments that don't fit in the previous bullets, please state them here. 


---

# Change working directory

* We are assuming you will store the notebooks in a subfolder, therefore when running the notebook in the editor, you will need to change the working directory

We need to change the working directory from its current folder to its parent folder
* We access the current directory with os.getcwd()

In [1]:
import os
current_dir = os.getcwd()
current_dir

'/workspace/Breast-Cancer-Prediction/jupyter_notebooks'

We want to make the parent of the current directory the new current directory
* os.path.dirname() gets the parent directory
* os.chir() defines the new current directory

In [2]:
os.chdir(os.path.dirname(current_dir))
print("You set a new current directory")

You set a new current directory


Confirm the new current directory

In [3]:
current_dir = os.getcwd()
current_dir

'/workspace/Breast-Cancer-Prediction'

---

# Getting Started: Load libraries and set options

In [4]:
#load libraries
import pandas as pd
import numpy as np 
import matplotlib.pyplot as plt
import seaborn as sns

# For Train , Test Spliting 
from sklearn.model_selection import train_test_split

# For Building Classifier Models
from sklearn.ensemble import RandomForestClassifier
from sklearn.linear_model import LogisticRegression
from sklearn.neighbors import KNeighborsClassifier

# For Evaluating the Accuray of the Classifiers 
from sklearn.metrics import classification_report , mean_squared_error , confusion_matrix
from sklearn import metrics

# For  Hyper parameter Tuning 
from sklearn.model_selection import RandomizedSearchCV
from sklearn.model_selection import GridSearchCV

# For Cross Validation
from sklearn.model_selection import KFold, StratifiedKFold,cross_val_score, cross_val_predict

In [5]:
df = pd.read_csv("outputs/datasets/cleaned/data.csv")

## Cross-validation and Hyper-parameter Tuning

#### Cross Validation Techniques 

#### K-Fold Validation

In [51]:
logit = LogisticRegression(max_iter=10000)

In [52]:
cv1 = KFold(n_splits= 13, random_state = 12, shuffle=True)   

scores_kfold_rfc = cross_val_score(RFC_model, X, Y, scoring='accuracy',cv=cv1, n_jobs=-1)
scores_kfold_lrc = cross_val_score(LR_model, X, Y, scoring='accuracy',cv=cv1, n_jobs=-1)
scores_kfold_knn = cross_val_score(KNNI, X, Y, scoring='accuracy',cv=cv1, n_jobs=-1)

print("The Kfold Cross Validation for our Random Forest Classifier yields %0.2f accuracy " % (scores_kfold_rfc.mean()))
print("The Kfold Cross Validation for our Logistic Regression Classifier yields %0.2f accuracy " % (scores_kfold_lrc.mean()))
print("The Kfold Cross Validation for our KNN Classifier yields %0.2f accuracy " % (scores_kfold_knn.mean()))



The Kfold Cross Validation for our Random Forest Classifier yields 0.96 accuracy 
The Kfold Cross Validation for our Logistic Regression Classifier yields 0.94 accuracy 
The Kfold Cross Validation for our KNN Classifier yields 0.93 accuracy 


/workspace/.pip-modules/lib/python3.8/site-packages/sklearn/linear_model/_logistic.py:458: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(
/workspace/.pip-modules/lib/python3.8/site-packages/sklearn/linear_model/_logistic.py:458: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-re

#### Stratified K-fold Validation

In [49]:
skfold = StratifiedKFold(n_splits=3, random_state=100, shuffle=True)
scores_skfold_ = cross_val_score(RFC_model, X, Y, scoring='accuracy', cv=skfold, n_jobs=-1)

scores_skfold_rfc = cross_val_score(RFC_model, X, Y, scoring='accuracy',cv=skfold , n_jobs=-1)
scores_skfold_lrc = cross_val_score(LR_model, X, Y, scoring='accuracy',cv=skfold , n_jobs=-1)
scores_skfold_knn = cross_val_score(KNNI, X, Y, scoring='accuracy',cv=skfold , n_jobs=-1)

print("The Kfold Cross Validation for our Random Forest Classifier yields %0.2f accuracy " % (scores_skfold_rfc.mean()))
print("The Kfold Cross Validation for our Logistic Regression Classifier yields %0.2f accuracy " % (scores_skfold_lrc.mean()))
print("The Kfold Cross Validation for our KNN Classifier yields %0.2f accuracy " % (scores_skfold_knn.mean()))

The Kfold Cross Validation for our Random Forest Classifier yields 0.96 accuracy 
The Kfold Cross Validation for our Logistic Regression Classifier yields 0.94 accuracy 
The Kfold Cross Validation for our KNN Classifier yields 0.93 accuracy 


/workspace/.pip-modules/lib/python3.8/site-packages/sklearn/linear_model/_logistic.py:458: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(
/workspace/.pip-modules/lib/python3.8/site-packages/sklearn/linear_model/_logistic.py:458: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-re

### Hyperparameters Tuning With RandomizedSearch CV

#### I. RandomizedSearch CV

This is the kind of search technique that moves around searching the best fit of combination in a random fashion for a fixed set of parameters over a predetermined iterations. This is the best search method incase of many numbers of hyperparameters as it reduces the computation cost. Therefore, as Random Forest Classifier depends on multiple hyperparaemters , we will be using RadomizedSearch CV for this.

The hyper parameter tuning is the process of evaluating the best fit hyperparameters set for our model. This is often refered to as search in Machine Learning models and can be categorized into two based on their best fit search patterns. 

In [50]:
X_train, X_test, Y_train, Y_test = train_test_split(X, Y, test_size=0.2, random_state=2)


random_search = {'criterion': ['entropy', 'gini'],
 'max_depth': list(np.linspace(5, 1200, 10, dtype = int)) + [None],
 'max_features': ['sqrt','log2', None],
 'min_samples_leaf': [4, 6, 8, 12],
 'min_samples_split': [3, 7, 10, 14],
 'n_estimators': list(np.linspace(5, 1200, 3, dtype = int))}

model = RandomForestClassifier()

modelrf = RandomizedSearchCV(estimator = model ,param_distributions = random_search, cv = 4, verbose= 5, random_state= 101, n_jobs = -1)

modelrf.fit(X_train, Y_train)

Fitting 4 folds for each of 10 candidates, totalling 40 fits
[CV 1/4] END criterion=gini, max_depth=5, max_features=sqrt, min_samples_leaf=6, min_samples_split=7, n_estimators=5;, score=0.921 total time=   0.0s
[CV 2/4] END criterion=gini, max_depth=5, max_features=sqrt, min_samples_leaf=6, min_samples_split=7, n_estimators=5;, score=0.956 total time=   0.0s
[CV 3/4] END criterion=gini, max_depth=5, max_features=sqrt, min_samples_leaf=6, min_samples_split=7, n_estimators=5;, score=0.947 total time=   0.0s
[CV 4/4] END criterion=gini, max_depth=5, max_features=sqrt, min_samples_leaf=6, min_samples_split=7, n_estimators=5;, score=0.929 total time=   0.0s
[CV 2/4] END criterion=entropy, max_depth=1200, max_features=log2, min_samples_leaf=6, min_samples_split=7, n_estimators=1200;, score=0.982 total time=   5.5s
[CV 1/4] END criterion=entropy, max_depth=536, max_features=sqrt, min_samples_leaf=6, min_samples_split=14, n_estimators=1200;, score=0.930 total time=   5.6s
[CV 1/4] END criterio

RandomizedSearchCV(cv=4, estimator=RandomForestClassifier(), n_jobs=-1,
                   param_distributions={'criterion': ['entropy', 'gini'],
                                        'max_depth': [5, 137, 270, 403, 536,
                                                      668, 801, 934, 1067, 1200,
                                                      None],
                                        'max_features': ['sqrt', 'log2', None],
                                        'min_samples_leaf': [4, 6, 8, 12],
                                        'min_samples_split': [3, 7, 10, 14],
                                        'n_estimators': [5, 602, 1200]},
                   random_state=101, verbose=5)

In [17]:
modelrf.best_params_

{'n_estimators': 1200,
 'min_samples_split': 7,
 'min_samples_leaf': 6,
 'max_features': 'log2',
 'max_depth': 1200,
 'criterion': 'entropy'}

In [18]:
Randomized_YPred =modelrf.predict(X_test)

print(classification_report(Randomized_YPred  , Y_test))

              precision    recall  f1-score   support

           0       0.94      0.96      0.95        68
           1       0.93      0.91      0.92        46

    accuracy                           0.94       114
   macro avg       0.94      0.93      0.94       114
weighted avg       0.94      0.94      0.94       114



In [19]:
metrics.accuracy_score(Y_test, Randomized_YPred)

0.9385964912280702

#### II. GridSearch CV

Next is the GridSearch CV that is used to find the best hyperparameters for our model based on a grid based search. This is implied in the KNN model to find the best fit value for k which is the hyper parameter for that module.

After performing the search the best fit hyperparameters for our random forest model are found to be 

In [20]:
X_train, X_test, Y_train, Y_test = train_test_split(X, Y, test_size=0.2, random_state=2)

params = {'n_neighbors' : list(range(1,10))}

KNN_cv= GridSearchCV(KNNI,params,cv=100)
KNN_cv.fit(X_train,Y_train)
Y_pred  = KNN_cv.predict(X_test)

In [21]:
k = KNN_cv.best_params_.get("n_neighbors")
print(f"The best fit value for k is {k}.")

The best fit value for k is 4.


In [22]:
KNN_cv.best_score_

0.9384999999999999

## Feature Selection

In [23]:
X = df.drop(['diagnosis'],axis =1)
Y = df["diagnosis"]

In [24]:
X.shape

(569, 30)

As the dimension of our data at the moment is 30, it is likely that the model we create is going to fall into the limitations of the curse of dimensionality. This can be solved with the use of a technique called Feature selection. This refers to the process of selecting a subset of attributes for efficient model construction.

#### Selecting The K Best Features for Our Dataset

In [25]:
selected_features = [] 

best_features = SelectKBest(chi2 , k = 5)


fit = best_features.fit(X , Y).get_support()

for bool, feature in zip(fit, df.columns):
     if bool:
        selected_features.append(feature)
print("The best features are:{}".format(selected_features)) # The list of your 5 best features

NameError: name 'SelectKBest' is not defined

#### Viewing The Scores of the Top-k Selected Features

In [26]:
feature_scores = {}

for key, value in zip(df.columns,best_features.fit(X , Y).scores_):
    feature_scores[key] = value


for key, value in feature_scores.items():
    if key in selected_features:
        print(f"{key} : {value}")

NameError: name 'best_features' is not defined

#### Dropping Unnecessary columns/features from the dataset.

In [27]:
features = df[["texture_mean" , "perimeter_mean" , "perimeter_se" , "texture_worst" , "perimeter_worst"]]
target = df["diagnosis"]

In [28]:
features.head()

texture_mean  perimeter_mean  perimeter_se  texture_worst  perimeter_worst
0         10.38          122.80         8.589          17.33           184.60
1         17.77          132.90         3.398          23.41           158.80
2         21.25          130.00         4.585          25.53           152.50
3         20.38           77.58         3.445          26.50            98.87
4         14.34          135.10         5.438          16.67           152.20

In [29]:
target.head()

0    1
1    1
2    1
3    1
4    1
Name: diagnosis, dtype: int64

In [30]:
assert features.shape[0] == target.shape[0]

### Final Model 

#### Implementing KNN in our selected features using the bestfit hyper-parameters 

In [31]:
X_train , X_test , Y_train , Y_test  = train_test_split(features , target , test_size = 0.33 , random_state = 42)

KNN_model = KNeighborsClassifier(n_neighbors = 4).fit(X_train , Y_train)

In [32]:
Y_pred = KNN_model.predict(X_test)
Y_pred

array([0, 1, 1, 0, 0, 1, 1, 1, 0, 0, 0, 1, 0, 0, 0, 1, 0, 0, 0, 1, 0, 0,
       1, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 1, 0, 1, 0, 0, 1, 0, 0,
       0, 0, 0, 0, 0, 0, 1, 1, 0, 0, 0, 0, 0, 1, 1, 0, 0, 1, 1, 0, 0, 0,
       1, 1, 0, 0, 1, 1, 0, 1, 0, 0, 0, 0, 0, 0, 1, 0, 0, 1, 1, 1, 0, 1,
       0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 0, 1, 1, 0, 1, 1, 0, 0, 0, 1, 0, 0,
       1, 0, 0, 1, 0, 1, 0, 0, 0, 1, 0, 0, 0, 1, 0, 1, 1, 0, 0, 1, 1, 1,
       0, 0, 0, 1, 0, 0, 0, 1, 0, 1, 0, 0, 1, 0, 1, 1, 1, 0, 1, 0, 0, 0,
       0, 1, 1, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 1, 0, 1, 0, 0, 0, 0, 0, 0,
       1, 0, 0, 0, 0, 0, 0, 1, 1, 0, 1, 0])

In [33]:
metrics.accuracy_score(Y_test, Y_pred)

0.9574468085106383

The accuracy of KNN model after hyperparameter tuning has increased.

#### Implementing RandomForest Classifier in our selected features using the bestfit hyper-parameters 

In [34]:
X_train , X_test , Y_train , Y_test  = train_test_split(features , target , test_size = 0.33 , random_state = 42)

model_rf = RandomForestClassifier()

RF_Model = RandomizedSearchCV(estimator = model_rf ,param_distributions = random_search, cv = 4, verbose= 5, random_state= 101, n_jobs = -1)

RF_Model.fit(X_train, Y_train)


Fitting 4 folds for each of 10 candidates, totalling 40 fits
[CV 2/4] END criterion=gini, max_depth=5, max_features=sqrt, min_samples_leaf=6, min_samples_split=7, n_estimators=5;, score=0.905 total time=   0.0s
[CV 3/4] END criterion=gini, max_depth=5, max_features=sqrt, min_samples_leaf=6, min_samples_split=7, n_estimators=5;, score=0.947 total time=   0.0s
[CV 4/4] END criterion=gini, max_depth=5, max_features=sqrt, min_samples_leaf=6, min_samples_split=7, n_estimators=5;, score=0.937 total time=   0.1s
[CV 1/4] END criterion=gini, max_depth=5, max_features=sqrt, min_samples_leaf=6, min_samples_split=7, n_estimators=5;, score=0.917 total time=   0.1s
[CV 1/4] END criterion=entropy, max_depth=1200, max_features=log2, min_samples_leaf=6, min_samples_split=7, n_estimators=1200;, score=0.948 total time=   8.9s
[CV 2/4] END criterion=entropy, max_depth=536, max_features=sqrt, min_samples_leaf=6, min_samples_split=14, n_estimators=1200;, score=0.916 total time=   8.9s
[CV 2/4] END criterio

RandomizedSearchCV(cv=4, estimator=RandomForestClassifier(), n_jobs=-1,
                   param_distributions={'criterion': ['entropy', 'gini'],
                                        'max_depth': [5, 137, 270, 403, 536,
                                                      668, 801, 934, 1067, 1200,
                                                      None],
                                        'max_features': ['sqrt', 'log2', None],
                                        'min_samples_leaf': [4, 6, 8, 12],
                                        'min_samples_split': [3, 7, 10, 14],
                                        'n_estimators': [5, 602, 1200]},
                   random_state=101, verbose=5)

In [35]:
Y_pred = RF_Model.predict(X_test)
metrics.accuracy_score(Y_test , Y_pred)

0.9521276595744681

The accuracy of Random Forest Classifier has also increased after hyperparameter tuning over the selected features.

## Conclusion

To sum up, initially a datset that consisted of the features or details related to the infected cells present in the breast were selected and understood better with the use of available python pandas library.  After that , the matplotlib and seaborn libraries were used to visually represent or map the relationship between various features present in our dataset. 

Now, after that the first primary models on classification algorithms such as Logistic Regression, KNN Classifier and Random Forest Classifier were built using the sklearn library and evaluated using the available metrics. 

After understanding the need of the cross validation for building these models, all of these primary models were rebuilt using the K-fold and Stratified Cross Validation techniques and likewise the accuracies were checked. Both the cross validation techniques , seemed to impact the accuracy of the model in an uniform manner with **Random Forest Classifier with the highest accuracy** here.

Next, was the turn to pick the best fit combination of hyperparameters for our Random Forest and KNN classifiers. Therefore, RandomizedSearch CV and GridSearch CV was used respectively to find the best combination of parameters for our models for better result. 

Derived Best Fit Hyper Parameters for Random Forest  : 

{'criterion': 'entropy', 'gini'        , <br>
 'max_depth': 5, 137, 270, 403, 536,  668, 801, 934, 1067, 1200, None , <br>
 'max_features': 'auto', 'sqrt', 'log2',   None, <br>
  'min_samples_leaf': 4, 6, 8, 12,<br>
 'min_samples_split': 3, 7, 10, 14,<br>
 'n_estimators': 5, 602, 1200 }
                
Derived Best Fit Hyper Parameters for KNN Classifier  : 

{'n_neighbors' : 4}

Finally after finding the best tuning hyper parameters for the model , the curse of dimensionality was taken into consideration and the SelectKbest feature selection technique was used to select top 5 best features from the dataset based on their scores. 

The best features were 'texture_mean', 'perimeter_mean', 'perimeter_se', 'texture_worst' and  'perimeter_worst'.

**Finally,** after the implementation of all these refinement techniques the primarily built models were rebult by plugging in the best fit hyperparameters and selected features. **Performing this demonstrated an increase in the accuracy of the models but both our classifiers that are Random Forest And KNN appear to hold the same amount of accuracy at the end.**

---

# Push files to Repo

* In case you don't need to push files to Repo, you may replace this section with "Conclusions and Next Steps" and state your conclusions and next steps.

In [ ]:
import os
try:
  # create here your folder
  # os.makedirs(name='')
except Exception as e:
  print(e)
